### DISCUSSION TITLE
Rocky - s-nail replaced mail - overriding action variable not working

### DISCUSSION THREAD

**Discussion by darmach** [fail2ban/fail2ban#3512] [Q&A] [answered]
Hello
I have a Rocky 9 in which s-nail replaced mail package.

S-nail mail interface does not support `mail -E 'set escape'` arguments, so sending mails throws errors, and they're sent without subjects. This can be fixed by removing `-E 'set escape '` from `mailcmd` var in `mail-whois-lines.conf` or `mail.conf` (whatever is used by the user)

I wanted to override `mailcmd` creating file: `actions.d/s-nail.local` like follows:
```
# action.d/s-nail.local extended "mail-whois-lines.conf":
# workarounds problems with s-nail mail interface implementation

[INCLUDES]
before = mail-whois-lines.conf

[Init]
# original:
## mailcmd = mail -E 'set escape' -s
# new:
mailcmd = mail -s
```

However this does not seem to work - if I manually change mailcmd to the same value it works.
Am I using INCLUDES correctly?

I will attempt to fix the root cause, but wanted to test such workaround/override first.

**Comment by sebres** [comment, marked as answer]
Hmm.... very strange. I can swear it must work. Basically there are many includes working in the similar way.
Did you restart the jails to update the configuration?
Anyway to see whether your customizations are correct you can check the dump, for example:
```
fail2ban-client --dp | grep 'mail '
```

The alternate ways to configure that without additional action file:
- do that in `mail-whois-lines.local` directly (instead of `s-nail.local`) without include section (local settings overwrite conf settings of the same file) and then use `mail-whois-lines` action.
- or supply this direct in jail.local as a parameter:
```ini
[DEFAULT]
mta = mail
action_mwl = %(known/action_mwl)s[mailcmd="mail -s"]
```
or directly action:
```ini
[DEFAULT]
action = %(action_)s
         mail-whois-lines[sender="%(sender)s", dest="%(destemail)s", logpath="%(logpath)s", chain="%(chain)s", mailcmd="mail -s"]
```

**Reply by darmach** [reply]
@sebres thank you for prompt response! I did restart fail2ban service - do I have to restart jail separately aswell?

fail2b-client --dp output confirms that my customization was not considered:
```
  [ ['actionstart', 'printf %b "Hi,\\n\nThe jail sshd has been started successfully.\\n\nRegards,\\n\nFail2Ban" | mail -E \'set escape\' -s "[Fail2Ban] sshd: started on <fq-hostname>" mymail@gmail.com'],
    ['actionstop', 'printf %b "Hi,\\n\nThe jail sshd has been stopped.\\n\nRegards,\\n\nFail2Ban" | mail -E \'set escape\' -s "[Fail2Ban] sshd: stopped on <fq-hostname>" mymail@gmail.com'],
    ['actionban', '( printf %b "Hi,\\n\nThe IP <ip> has just been banned by Fail2Ban after\n<failures> attempts against sshd.\\n\\n\nHere is more information about <ip> :\\n"\nwhois <ip> || echo "missing whois program";\nprintf %b "\\nLines containing failures of <ip> (max 1000)\\n";\nlogpath="/var/log/secure"; grep -m 1000 -wF "<ip>" $logpath | tail -n 1000;\nprintf %b "\\n\nRegards,\\n\nFail2Ban" ) | mail -E \'set escape\' -s "[Fail2Ban] sshd: banned <ip> from <fq-hostname>" mymail@gmail.com'],
    ['mailcmd', "mail -E 'set escape' -s"]]]
```

I'll try other options this afternoon. While we're talking - maybe I could open a PR to get rid of `-E 'set escape' ` completely? What is it's purpose in this case?

**Reply by sebres** [reply]
> do I have to restart jail separately aswell?

No.

> fail2b-client --dp output confirms that my customization was not considered

Strange. I tried it right now (exactly like you did) and see `mail -s` everywhere. Which fail2ban version do you have?
And did you add the action `s-nail` to the jails actions, don't you?

> I could open a PR to get rid of -E 'set escape'  completely?

There are already an issue for that #3059 and RFE #3071 

> What is it's purpose in this case?

This was a fix for theoretical RCE vulnerability CVE-2021-32749 (see [GHSA-m985-3f3v-cwmm](https://github.com/fail2ban/fail2ban/security/advisories/GHSA-m985-3f3v-cwmm)).
I was listening that upstream mailutils fixed the behavior to disable escapes when STDIN is not a TTY, just not sure it'd take place for every distribution across the world, thus it still unchanged.

I never understand the necessity to mail every ban (for instance on my hosts I'd get many thousands messages per day from each machine), let alone the mailing actions like that may extremely delay the banning process (if mail/whois/whatever by previous ban is too slow or even hang up to the timeout due to some network issue or load/quota on whois database).
This is totally unneeded, ugly and simply bad practice, so I know nobody who uses that in production in earnest.

**Reply by darmach** [reply]
Oh, yes, I've seen you commenting about mails for every ban somewhere around., I'm running 1.0.2-3 from epel repos on RL9.

I think it all boils down to use cases. In typical internet-facing infrastructure, indeed there is not much sense for spamming the mails (because of amount of it), however if you consider small scale, or critical ("almost" closed off deployments - like mine, another home machine, but of course very valuable to me) it's good to know that there were some attempts. This means that something really bad happened, if someone would get that far to be able to attempt a logon to this machine.

I am setting it up, while expecting to NEVER get this email ;) 

As for adding the s-nail action to my jail, I'll let you guess....  :D  ...so we can consider that one mystery solved! Sorry to waste your time on a problem between the chair and the keyboard!
